Option improvements

rocky opened the pull request:
Not totally sure about this one. I at least wanted better error reporting on giving a bad option. 

I had considered putting all of the patches into a HashMap where the keys are the patch option and the  values contain the particular VM parametrized by the specific patch. But this would mean creating all of the combinations up front, even though one of them is used. 

A midway approach is to list all of the valid patches in a vector that can be displayed in both help and on error. What do you think?

rocky committed:
WIP option improvements

whilei commented:
I guess I would lean toward the "midway" option... (since a VM _could_ also use a trait object with multiple patches...)

But for the "hashmap" idea -- although it seems kind of wasteful, I guess the actual overhead would be pretty low if it would only need to be created once, or?

Just another idea on the table - while on one hand I like the idea of keeping configuration documentation "close to code" (where available options on error would be, like, printed to the terminal), it also seems kind of redundant to create extra variables just for logging configuration errors when available configuration is already auto-documented thru Rust, ala https://docs.rs/sputnikvm/0.10.1/sputnikvm/struct.EmbeddedByzantiumPatch.html. Might there be a clean way for an error to simply print a relevant URL for reference? Just an idea.

rocky commented:
> But for the "hashmap" idea -- although it seems kind of wasteful, I guess the actual overhead would be pretty low if it would only need to be created once, 

It occurs to me that how you'd do this in other programming languages is just store a function pointer and then instantiate or call the "new" only after seeing the option. My Rust knowledge just isn't up to speed yet to do that. 

> Might there be a clean way for an error to simply print a relevant URL for reference? Just an idea.

Sure, that makes sense. I need to put my work on this on hold for a little bit, but I'll do that if I come back to this.

rocky committed:
Put patch options in a HashMap...

code is a little more DRY. Also we can show valid patch names by listing
the keys of map.

rocky committed:
Describe what it takes to not pass an iterator

BuHeP commented on cli/src/bin/main.rs:
```diff
@@ -15,14 +15,67 @@ use std::fs::File;
 use profiler::Profiler;
 use bigint::{Gas, Address, U256, M256, H256};
 use hexutil::read_hex;
-use sputnikvm::{HeaderParams, Context, SeqTransactionVM, ValidTransaction, VM,
```
18

BuHeP replied:
- [ 18 ] 

- [ ] 

- [ ] 

- [ ] 

- [ ] 

- [ ]
